grpc/oob/diagnostic/screenshot.go

```diff
@@ -0,0 +1,138 @@
+package diagnostic
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/bmc-toolbox/bmclib/v2"
+	"github.com/bmc-toolbox/bmclib/v2/bmc"
+	"github.com/bmc-toolbox/bmclib/v2/providers"
+	"github.com/go-logr/logr"
+	"github.com/prometheus/client_golang/prometheus"
+	v1 "github.com/tinkerbell/pbnj/api/v1"
+	common "github.com/tinkerbell/pbnj/grpc/oob"
+	"github.com/tinkerbell/pbnj/pkg/metrics"
+	"github.com/tinkerbell/pbnj/pkg/repository"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type Action struct {
+	common.Accessory
+	ScreenshotRequest *v1.ScreenshotRequest
+}
+
+// WithLogger adds a logr to an Action struct.
+func WithLogger(l logr.Logger) Option {
+	return func(a *Action) error {
+		a.Log = l
+		return nil
+	}
+}
+
+// Option to add to an Actions.
+type Option func(a *Action) error
+
+func NewScreenshotter(req *v1.ScreenshotRequest, opts ...Option) (*Action, error) {
+	a := &Action{}
+	a.ScreenshotRequest = req
+	for _, opt := range opts {
+		err := opt(a)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return a, nil
+
+}
+
+func (m Action) GetScreenshot(ctx context.Context) (image []byte, filetype string, err error) {
+	labels := prometheus.Labels{
+		"service": "diagnostic",
+		"action":  "screenshot",
+	}
+
+	timer := prometheus.NewTimer(metrics.ActionDuration.With(labels))
+	defer timer.ObserveDuration()
+
+	tracer := otel.Tracer("pbnj")
+	ctx, span := tracer.Start(ctx, "diagnostic.GetScreenshot", trace.WithAttributes(
+		attribute.String("bmc.device", m.ScreenshotRequest.GetAuthn().GetDirectAuthn().GetHost().GetHost()),
+	))
+	defer span.End()
+
+	if v := m.ScreenshotRequest.GetVendor(); v != nil {
+		span.SetAttributes(attribute.String("bmc.vendor", v.GetName()))
+	}
+
+	host, user, password, parseErr := m.ParseAuth(m.ScreenshotRequest.GetAuthn())
+	if parseErr != nil {
+		span.SetStatus(codes.Error, "error parsing credentials: "+parseErr.Error())
+		return nil, "", parseErr
+	}
+	span.SetAttributes(attribute.String("bmc.host", host), attribute.String("bmc.username", user))
+
+	opts := []bmclib.Option{
+		bmclib.WithLogger(m.Log),
+		bmclib.WithPerProviderTimeout(45 * time.Second),
```

Comment:
Theres a helper method here to retrieve the timeout set through env vars,

https://github.com/tinkerbell/pbnj/blob/629921c5f6e4282ef1a2281a8995e5801b0ddeaa/grpc/oob/common.go#L118